### ISSUE TITLE
Implementing Arena up to 1.2: Verifying Activations

### ISSUE THREAD

**Issue by BlastWind** [BlastWind/hech-interp#11]
With cache activations implemented, we can begin implementing the ARENA. For starters, we implement up to (and including) the "verify activations" exercise from the ARENA 1.2 [notebook](https://colab.research.google.com/github/callummcdougall/ARENA_3.0/blob/main/chapter1_transformer_interp/exercises/part2_intro_to_mech_interp/1.2_Intro_to_Mech_Interp_solutions.ipynb?t=20250211#scrollTo=maKoYy0IpQVu).

Here are the python code to support:
```python

# 1. User specifies device
device = torch.device("mps" if t.backends.mps.is_available() else "cuda" if t.cuda.is_available() else "cpu")

# 2. Turn off gradient to speed up
torch.set_grad_enabled(False)

# 3. Loading a model
gpt2_small: HookedTransformer = HookedTransformer.from_pretrained("gpt2-small")

# 4. Inferencing and getting loss
loss = gpt2_small("Some random text chunk we want GPT2 to complete", return_type="loss")

# 5. Token and string conversions
gpt2.to_str_tokens("gpt2")            # ['<|endoftext|>', 'g', 'pt', '2']
gpt2.to_str_tokens(["gpt2", "gpt2"])  # [['<|endoftext|>', 'g', 'pt', '2'], ['<|endoftext|>', 'g', 'pt', '2']]
gpt2.to_tokens("gpt2")                # tensor([[50256,    70,   457,    17]], device='cuda:0')
gpt2.to_string([50256, 70, 457, 17])  # <|endoftext|>gpt2

# 6. Running with activation cache
gpt2_text = "Natural language processing tasks, such as question answering, machine translation, reading comprehension, and summarization, are typically approached with supervised learning on taskspecific datasets."
gpt2_tokens = gpt2_small.to_tokens(gpt2_text)
gpt2_logits, gpt2_cache = gpt2_small.run_with_cache(gpt2_tokens, remove_batch_dim=True)

# 7. Verifying activations (a quick check that we can calculate `pattern` from `q, k`)
layer0_pattern_from_cache = gpt2_cache["pattern", 0]

q, k = gpt2_cache["q", 0], gpt2_cache["k", 0]
seq, nhead, headsize = q.shape
layer0_attn_scores = einops.einsum(q, k, "seqQ n h, seqK n h -> n seqQ seqK")
mask = t.triu(t.ones((seq, seq), dtype=t.bool), diagonal=1).to(device)
layer0_attn_scores.masked_fill_(mask, -1e9)
layer0_pattern_from_q_and_k = (layer0_attn_scores / headsize**0.5).softmax(-1)

t.testing.assert_close(layer0_pattern_from_cache, layer0_pattern_from_q_and_k)
print("Tests passed!")
```

**Comment by BlastWind** [comment]
The implementation order, I think, should follow:
1. Loading a model (this should already be mostly done)
2. Inferencing, no loss for now (this should be done)
3. Token and string conversion functions
4. Caching a run (this should be done)
5. Verifying activations
6. Track loss in run
7. Toggable grad
8. Specifiable device
